nmigen/vendor/quicklogic.py

```diff
@@ -0,0 +1,214 @@
+
+        assert toolchain in ("Quicklogic")
+        self.toolchain = toolchain
+
+    @property
+    def required_tools(self):
+        if self.toolchain == "Quicklogic":
+            return self._quicklogic_required_tools
+        assert False
+
+    @property
+    def file_templates(self):
+        if self.toolchain == "Quicklogic":
+            return self._quicklogic_file_templates
+        assert False
+
+    @property
+    def command_templates(self):
+        if self.toolchain == "Quicklogic":
+            return self._quicklogic_command_templates
+        assert False
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "TRUE"
+
+    def get_input(self, pin, port, attrs, invert):
+        self._check_feature("single-ended input", pin, attrs,
+                            valid_xdrs=(0, 1, 2), valid_attrs=True)
+        m = Module()
+        m.d.comb += pin.i.eq(self._invert_if(invert, port))
+        return m
+
+    def get_output(self, pin, port, attrs, invert):
+        self._check_feature("single-ended output", pin, attrs,
+                            valid_xdrs=(0, 1, 2), valid_attrs=True)
+        m = Module()
+        i, o, t = self._get_xdr_buffer(m, pin, o_invert=invert)
+        m.d.comb += port.eq(self._invert_if(invert, o))
+        return m
+
+    def _get_xdr_buffer(self, m, pin, *, i_invert=False, o_invert=False):
+        def get_ineg(y, invert):
+            if invert:
+                a = Signal.like(y, name_suffix="_n")
+                m.d.comb += y.eq(~a)
+                return a
+            else:
+                return y
+
+        def get_oneg(a, invert):
+            if invert:
+                y = Signal.like(a, name_suffix="_n")
+                m.d.comb += y.eq(~a)
+                return y
+            else:
+                return a
+
+        if "i" in pin.dir:
+            if pin.xdr < 2:
+                pin_i  = get_ineg(pin.i,  i_invert)
+            elif pin.xdr == 2:
+                pin_i0 = get_ineg(pin.i0, i_invert)
+                pin_i1 = get_ineg(pin.i1, i_invert)
+        if "o" in pin.dir:
+            if pin.xdr < 2:
+                pin_o  = get_oneg(pin.o,  o_invert)
+            elif pin.xdr == 2:
+                pin_o0 = get_oneg(pin.o0, o_invert)
+                pin_o1 = get_oneg(pin.o1, o_invert)
+
+        i = o = t = None
+        if "i" in pin.dir:
+            i = Signal(pin.width, name="{}_xdr_i".format(pin.name))
+        if "o" in pin.dir:
+            o = Signal(pin.width, name="{}_xdr_o".format(pin.name))
+        if pin.dir in ("oe", "io"):
+            t = Signal(1,         name="{}_xdr_t".format(pin.name))
+
+        if pin.xdr == 0:
+            if "i" in pin.dir:
+                i = pin_i
+            if "o" in pin.dir:
+                o = pin_o
+            if pin.dir in ("oe", "io"):
+                t = ~pin.oe
+        else:
+            assert False
```

Comment:
Does Quicklogic have registered and DDR inputs? If yes, they should be supported. If not, they should not be advertised above.

Reply:
Neither SDR nor DDR are supported. I set `valid_xdrs=0` and I [simplified](https://github.com/nmigen/nmigen/pull/504/files#diff-5b82fd62f94e74b0762f9965e2b93236426be4957ebbaa42325de8c0501ec880R172-R175) handling them a bit since there is only one option.

Reply:
Okay, but now this code doesn't make any sense. As far as I can tell, you can just remove `get_input` and `get_output` since they are not doing anything useful.

Reply:
All right - removed them

Reply:
You should then also remove `_get_xdr_buffer`, since it is not used anymore. Please put some effort into addressing review comments, because it is nearing the point where it would be easier for me to rewrite your code than to review it.